File: MultiLanguageSandbox/src/test/deps/perl/33.pl

sub max_herb_value {
    my ($t, $herbs) = @_;
    
    # Initialize a DP array where dp[i] is the max value for time i
    my @dp;
    $dp[0] = 0;  # Base case: 0 time gives 0 value
    
    for my $i (1..$t) {
        $dp[$i] = 0;
        foreach my $herb (@$herbs) {
            my ($time, $value) = @$herb;
            if ($time <= $i) {
                if ($dp[$i - $time] + $value > $dp[$i]) {
                    $dp[$i] = $dp[$i - $time] + $value;
                }
            }
        }
    }
    
    return $dp[$t];
}


sub test_max_herb_value {
die "Test failed!" unless max_herb_value(10, [[1, 5], [2, 10], [3, 15]]) == 50;
die "Test failed!" unless max_herb_value(5, [[2, 4], [3, 5]]) == 9;
die "Test failed!" unless max_herb_value(7, [[1, 2], [2, 5], [3, 7]]) == 17;
die "Test failed!" if max_herb_value(0, [[1, 10], [2, 20]]) != 0;
die "Test failed!" unless max_herb_value(20, [[5, 10], [4, 7], [3, 5]]) == 40;
print "All tests passed!\n";
}

test_max_herb_value();

File: MultiLanguageSandbox/src/test/deps/ruby/33.rb


# Generates all unique four-digit permutations using the given four digits.
# Each digit is used exactly once in each permutation. The function accepts
# four integers as arguments and returns an array of strings, each representing
# a unique permutation. The permutations are sorted in ascending order.
#
# Example:
# >>> generate_four_digit_permutations(1, 2, 3, 4)
# ["1234", "1243", "1324", "1342", "1423", "1432", ...]
# >>> generate_four_digit_permutations(1, 2, 3, 5)
# ["1235", "1253", "1325", "1352", "1523", "1532", ...]


def generate_four_digit_permutations(*digits)
  # Ensure we have exactly four digits, "Exactly four digits must be provided" unless digits.size == 4

  # Generate all permutations of the four digits
  permutations = digits.permutation.to_a

  # Convert each permutation array to a string
  result = permutations.map { |perm| perm.join }

  # Sort the result in ascending order
  result.sort
end

raise 'Test failed' unless generate_four_digit_permutations(1, 2, 3, 4) == ["1234", "1243", "1324", "1342", "1423", "1432", "2134", "2143", "2314", "2341", "2413", "2431", "3124", "3142", "3214", "3241", "3412", "3421", "4123", "4132", "4213", "4231", "4312", "4321"]

raise 'Test failed' unless generate_four_digit_permutations(1, 2, 3, 5) == ["1235", "1253", "1325", "1352", "1523", "1532", "2135", "2153", "2315", "2351", "2513", "2531", "3125", "3152", "3215", "3251", "3512", "3521", "5123", "5132", "5213", "5231", "5312", "5321"]
   


puts 'All tests passed!'